crates/snark-acct-sys/src/verification.rs

```diff
@@ -0,0 +1,197 @@
+use ssz::Encode as _;
+use strata_acct_types::{AccountId, AcctError, AcctResult, BitcoinAmount, Mmr64, StrataHasher};
+use strata_ledger_types::{ISnarkAccountState, ISnarkAccountStateExt, IStateAccessor};
+use strata_merkle::{MerkleProof, hasher::MerkleHasher};
+use strata_snark_acct_types::{
+    LedgerRefProofs, MessageEntryProof, ProofState, SnarkAccountUpdate,
+    SnarkAccountUpdateContainer, UpdateOperationData, UpdateOutputs, UpdateProofPubParams,
+};
+
+/// Verifies an account update is correct with respect to the current state of
+/// snark account, including checking account balances.
+pub fn verify_update_correctness<'a, S: IStateAccessor>(
+    state_accessor: &S,
+    sender: AccountId,
+    snark_state: &impl ISnarkAccountState,
+    update: &'a SnarkAccountUpdateContainer,
+    cur_balance: BitcoinAmount,
+) -> AcctResult<VerifiedUpdate<'a>> {
+    let operation = update.base_update().operation();
+    let outputs = operation.outputs();
+
+    // 1. Check seq_no matches
+    let expected_seq = snark_state.seqno().incr();
+    if operation.seq_no() != *expected_seq.inner() {
+        return Err(AcctError::InvalidUpdateSequence {
+            account_id: sender,
+            expected: *expected_seq.inner(),
+            got: operation.seq_no(),
+        });
+    }
+
+    // 2. Check message counts / proof indices line up
+    let cur_idx = snark_state.get_next_inbox_msg_idx();
+    let new_idx = operation.new_state().next_inbox_msg_idx();
+    let processed_len = operation.processed_messages().len() as u64;
+
+    let expected_idx = cur_idx
+        .checked_add(processed_len)
+        .ok_or(AcctError::MsgIndexOverflow { account_id: sender })?;
+    if expected_idx != new_idx {
+        return Err(AcctError::InvalidMsgIndex {
+            account_id: sender,
+            expected: expected_idx,
+            got: new_idx,
+        });
+    }
+
+    let accum_proofs = update.accumulator_proofs();
+    // 3. Verify ledger references using the provided state accessor
+    verify_ledger_refs(
+        sender,
+        state_accessor.asm_manifests_mmr(),
+        accum_proofs.ledger_ref_proofs(),
+    )?;
+
+    // 4. Verify input mmr proofs
+    verify_input_mmr_proofs(sender, snark_state, accum_proofs.inbox_proofs())?;
+
+    // 4. Verify outputs can be applied safely
+    verify_update_outputs_safe(outputs, state_accessor, cur_balance)?;
+
+    // 5. Verify the witness check
+    verify_update_witness(sender, snark_state, update.base_update())?;
+
+    Ok(VerifiedUpdate { operation })
+}
+
+fn verify_ledger_refs(
+    account: AccountId,
+    mmr: &Mmr64,
+    ledger_ref_proofs: &LedgerRefProofs,
+) -> AcctResult<()> {
+    let generic_mmr = mmr.to_generic();
+    for proof in ledger_ref_proofs.l1_headers_proofs() {
+        let hash = proof.entry_hash();
+        let cohashes: Vec<[u8; 32]> = proof.proof().cohashes().to_vec();
+        let generic_proof = MerkleProof::from_cohashes(cohashes, proof.entry_idx());
```

Comment:
This is doing an extra copy of all of the proof node hashes here.  It should be possible to verify it directly by ref, possibly needing some extra conversions.